reason/guided_search/rstar.py

```diff
@@ -0,0 +1,200 @@
+                # expand its children through legal_action_update,
+                # for select, if there is unexplored children, select it randomly. if all children nodes
+                # have been explored, select UCB
+                next_node, is_leaf = self._select_child(node, env_copy, i_path)     # find a leaf node or
+                                                                            # simulate the remaining
+                node_path.append(next_node)
+
+                done = env_copy.is_terminal(
+                    next_node
+                )       # checking terminal condition
+                # update legal action (expand) when the current code is not a leaf (no children)
+                # no env.step only use for checking termination when is not leaf, and update legal action
+                # when the current node is leaf
+                # print(f"Path {i_path}: depth = {next_node.depth}, done = {done}, is leaf = {is_leaf}")
+
+                if not done and is_leaf:     # expand when encounter a leaf node one step further
+                    next_node_children = env_copy.try_update_legal_action(node = next_node)
+                    for c in next_node_children:
+                        c.set_rollout_id(i_path)
+                    self.parent2children[next_node]=next_node_children
+
+                if self.show_tree_expansion:
+                    self.draw_tree()
+
+                if done:
+                    self.explored_nodes.add(next_node)
+
+                node=next_node
+
+            else:
+                # boostrapping
+                reward = next_node.calculate_reward()
+
+                for node in reversed(node_path):
+                    self.Q[node] += reward
+                    self.N[node] += 1
+                    self.explored_nodes.add(node)
+
+            model_rollout_nodes.append(next_node)
+            _, best_solution, _, chosen_node, all_solution_nodes, all_solutions = stochastic_find_best_solution(
+                self.root, env_copy.evaluator, enable_potential_score=False
+            )
+
+            # model_solutions.append(best_solution)
+            # model_all_solutions.append(all_solutions)
+            assert best_solution is not None
+
+            traj_data = {
+                "path_idx": i_path,
+                "text": best_solution,
+                "value": reward,
+                "api_completion_tokens": env_copy.total_api_call_completion,
+                "tree_completion_tokens": env_copy.total_tree_completion,
+            }
+
+            traj_list.append(traj_data)
+
+        return traj_list
+
+    def draw_tree(self):
+
+        def display_tree(node):
+            print("|" + "-" * (node.depth * 4) + str(node))
+            for child in node.children:
+                display_tree(child)
+
+        print(f"\n---------Expanded Tree---------")
+        display_tree(self.root)
```

Comment:
this function is useful for both `RStartSearchTree` and `SearchTree`, consider moving into `SearchTree`

I can do this, what's your opinion @YanSong97

Reply:
yeah looks good to me